cpp/examples/src/example_client_perf_test.cpp

```diff
@@ -0,0 +1,131 @@
+#include <foxglove/websocket/websocket_client.hpp>
+#include <foxglove/websocket/websocket_notls.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <iostream>
+#include <memory>
+#include <regex>
+#include <unordered_map>
+
+struct ChannelStats {
+  std::string topic;
+  size_t bytesReceived = 0;
+  size_t msgsReceived = 0;
+  std::chrono::time_point<std::chrono::steady_clock> firstMsgTime;
+  std::chrono::time_point<std::chrono::steady_clock> lastMsgTime;
+};
+
+constexpr char DEFAULT_URI[] = "ws://localhost:8765";
+constexpr auto DEFAULT_TIMEOUT = std::chrono::seconds(5);
+
+std::atomic<bool> doAbort = false;
+
+void signal_handler(int) {
+  doAbort = true;
+}
+
+int main(int argc, char** argv) {
+  const auto url = argc > 1 ? argv[1] : DEFAULT_URI;
+  const auto topicRegex = argc > 2 ? std::regex(argv[2]) : std::regex(".*");
+  std::unordered_map<foxglove::SubscriptionId, ChannelStats> channelStatsBySubId;
+  std::unordered_map<foxglove::ChannelId, foxglove::SubscriptionId> subIdByChannelId;
+  foxglove::SubscriptionId nextSubId = 0;
+  foxglove::Client<foxglove::WebSocketNoTls> client;
+  size_t maxTopicCharLength = 0;
+
+  client.setBinaryMessageHandler([&](const uint8_t* data, size_t dataLength) {
+    if (static_cast<foxglove::BinaryOpcode>(data[0]) != foxglove::BinaryOpcode::MESSAGE_DATA) {
+      return;
+    }
```

Comment:
Drive-by API review comment: this generic `set(Text|Binary)MessageHandler` API for the Client is not great – custom client code should not be responsible for parsing the `op`/`BinaryOpcode` itself. The point of providing a library for people to use is that it should take care of these protocol details itself so the user doesn't have to.

Reply:
The websocket client was primarily added for use in smoke tests, so the API is rather low level. But in general I agree with you.